neothesia/src/scene/playing_scene/midi_player.rs

```diff
@@ -198,69 +251,202 @@ pub enum MidiEventSource {
 struct UserPress {
     timestamp: Instant,
     note_id: u8,
+    time_key_up: Option<Instant>,
+    occurrence: usize,
 }
 
-#[derive(Debug)]
 pub struct PlayAlong {
     user_keyboard_range: piano_math::KeyboardRange,
 
-    required_notes: HashSet<u8>,
-
+    required_notes: Vec<UserPress>,
+    finished: bool,
     // List of user key press events that happened in last 500ms,
     // used for play along leeway logic
-    user_pressed_recently: VecDeque<UserPress>,
+    user_pressed_recently: Vec<UserPress>,
+    user_stats: NoteStats, // struct to finalize the stats log
+
+    user_notes: Vec<UserPress>,     // log all user notes to get durrations
+    file_notes: Vec<UserPress>,     // log all file notes to compare against user
+    occurrence: HashMap<u8, usize>, // Keeping user to file log incremental pointer rewind immune
+    on_finish: Option<Box<dyn Fn()>>,
 }
 
 impl PlayAlong {
-    fn new(user_keyboard_range: piano_math::KeyboardRange) -> Self {
+    fn new(user_keyboard_range: piano_math::KeyboardRange, user_stats: NoteStats) -> Self {
         Self {
             user_keyboard_range,
             required_notes: Default::default(),
             user_pressed_recently: Default::default(),
+
+            occurrence: HashMap::new(),
+            finished: Default::default(),
+
+            user_notes: Default::default(),
+            file_notes: Default::default(),
+            on_finish: None,
+            user_stats,
         }
     }
+    pub fn set_on_finish<F>(&mut self, callback: F)
+    where
+        F: Fn() + 'static,
+    {
+        self.on_finish = Some(Box::new(callback));
+    }
 
     fn update(&mut self) {
-        // Instead of calling .elapsed() per item let's fetch `now` once, and subtract it ourselves
         let now = Instant::now();
+        let threshold = Duration::from_millis(500);
 
-        while let Some(item) = self.user_pressed_recently.front_mut() {
+        // Track the count of items before retain
+        let count_before = self.user_pressed_recently.len();
+
+        // Retain only the items that are within the threshold
+        self.user_pressed_recently.retain(|item| {
             let elapsed = now - item.timestamp;
+            elapsed <= threshold
+        });
 
-            // If older than 500ms
-            if elapsed.as_millis() > 500 {
-                self.user_pressed_recently.pop_front();
-            } else {
-                // All subsequent items will by younger than front item, so we can break
-                break;
-            }
+        // Calculate the count of deleted items
+        let count_deleted = count_before - self.user_pressed_recently.len();
+        if count_deleted > 0 {
+            self.user_stats.wrong_notes += count_deleted as i32;
         }
```

Comment:
both `count_before` and `len()` are unsigned values therefore `count_deleted` also is, they can't contain negative values (program will just crache in debug mode if that happens), so either you have to guard agains this value going negative (as it will overflow the intager), or if it can't happen just drop the if, as it does nothing in it's current form.

Reply:
It turns out that logic has been removed 3 days ago in this commit of the after-tests debugging:

https://github.com/PolyMeilex/Neothesia/commit/2379e399505ded40fbacb39c7038aac64e9b2eec

doing 

self.user_stats.wrong_notes -= 1;

That, was a stupid idea to begin with. Everything in update() should be always only wrong notes, good ones shouldn't be left inside the collection to expire.

So the rest was just left-over dead code, i cleaned it.

Test done to validate good notes are not left in update() as wrong notes: 

1. Let a note sit down for more than 500ms, press it after 2 3 seconds as a slow hit, its never a "Wrong note" must be a "slow hit"